numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -1086,6 +1086,95 @@ TIMEDELTA_setitem(PyObject *op, void *ov, void *vap)
 
 /* Assumes contiguous, and aligned, from and to */
 
+/* special case TIMEDELTA NPY_DATETIME_NAT handling */
```

Comment:
Note that a large amount of code duplication is introduced because of this specialization, but I think the choice is always going to be between highly-specialized loops for type-type conversion & more general loops where bystander type conversions suffer a performance hit from NaT checking.